### PR TITLE
[Enhancement] support privilege for pipe

### PR DESCRIPTION
**Pull request by murphyatwork**
- Make `PIPE` as an object inside a database, users need `CREATE PIPE` privilege of database to create it
- Privilege types of pipe: `ALTER` to `ALTER PIPE`, `DROP` to `DROP PIPE`, `USAGE` to `SHOW PIPES` and `DESC PIPE`

Some example statements:
```sql
grant USAGE on PIPE db1.p1 to test;
grant DROP on PIPE db1.p1 to test;
grant ALTER on PIPE p1 to test;

grant USAGE on ALL PIPES in DATABASE db1 to test;
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -1615,18 +1615,21 @@ privilegeType
...
+        DATABASE| TABLE| VIEW| FUNCTION| GLOBAL FUNCTION| MATERIALIZED VIEW|
+        RESOURCE| RESOURCE GROUP| EXTERNAL CATALOG | POLICY | STORAGE VOLUME
+        | PIPE )
     | DELETE | DROP | EXPORT | FILE | IMPERSONATE | INSERT | GRANT | NODE | OPERATE
     | PLUGIN | REPOSITORY| REFRESH | SELECT | UPDATE | USAGE
     ;
 
 privObjectType
     : CATALOG | DATABASE | MATERIALIZED VIEW | POLICY | RESOURCE | RESOURCE GROUP| STORAGE VOLUME | SYSTEM | TABLE| VIEW
+    | PIPE
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1627]: pipe is reserved?

File: fe/fe-core/src/main/java/com/starrocks/privilege/PipePEntryObject.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    @Override
+    public boolean match(Object obj) {
+        if (!(obj instanceof PipePEntryObject)) {
+            return false;
+        }
+        PipePEntryObject other = (PipePEntryObject) obj;
+        if (Objects.equals(other.getDbUUID(), PrivilegeBuiltinConstants.ALL_DATABASES_UUID)) {
+            return true;
+            //            return Objects.equals(getName(), other.getName());
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PipePEntryObject.java:91]:
```suggestion
```

**Review by HangyuanLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 1 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1627]: it's not.

File: fe/fe-core/src/main/java/com/starrocks/privilege/ranger/starrocks/RangerStarRocksAccessController.java [hunk 3]
```diff
@@ -270,6 +271,22 @@ public void checkResourceGroupAction(UserIdentity currentUser, Set<Long> roleIds
         hasPermission(resource, currentUser, privilegeType);
     }
 
+    @Override
+    public void checkPipeAction(UserIdentity currentUser, Set<Long> roleIds, PipeName name, PrivilegeType privilegeType)
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/ranger/starrocks/RangerStarRocksAccessController.java:275]: also need to update the sevice-def.json? @HangyuanLiu

File: fe/fe-core/src/test/java/com/starrocks/sql/analyzer/PrivilegeCheckerTest.java [hunk 4]
```diff
@@ -3049,6 +3050,162 @@ public void testStorageVolumeStatement() throws Exception {
                 "Access denied;");
     }
 
+    @Test
+    public void testPipe() throws Exception {
```
Comment (nshangyiming) [inline fe/fe-core/src/test/java/com/starrocks/sql/analyzer/PrivilegeCheckerTest.java:3054]: need to add more ut, e.g., show grants for pipe obj is not tested

**Review by nshangyiming — COMMENTED** [review]
PTAL

File: fe/fe-core/src/test/java/com/starrocks/sql/analyzer/PrivilegeCheckerTest.java (hunk 4 above)
Comment (nshangyiming) [reply fe/fe-core/src/test/java/com/starrocks/sql/analyzer/PrivilegeCheckerTest.java:3054]: and also select from information_schema.grants_to_user, grants_to_role, better to use sql tests to cover it, ut is not easy

Comment (mofeiatwork) [reply fe/fe-core/src/test/java/com/starrocks/sql/analyzer/PrivilegeCheckerTest.java:3054]: the `information_schema.grants*` should be same as `show grants`, is it necessary to test it again ?

**Review by HangyuanLiu — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**